### PR TITLE
API docs for Invariant, Functor, Applicative and Monad

### LINKED ISSUE #147: Invariant API Docs
## Description

Add API docs for `Invariant`.

### LINKED ISSUE #148: Functor API docs
## Description

Add API docs for Functor.

### LINKED ISSUE #149: Applicative API docs
## Description

Add API docs for `Applicative`.

### LINKED ISSUE #150: Monad API docs
## Description

Add API docs for `Monad`.

### PR DESCRIPTION
**Pull request by truizlop**
## Related issues

- Closes #147.
- Closes #148.
- Closes #149.
- Closes #150.

## Goal

Add API docs for `Invariant`, `Functor`, `Applicative` and `Monad`.

### CODE REVIEW

File: Sources/Bow/Typeclasses/Applicative.swift [hunk 1]
```diff
@@ -1,77 +1,199 @@
 import Foundation
 
+/// An Applicative Functor is a `Functor` that also provides functionality to embed pure expressions, and sequence computations and combine their results.
```
Comment (pepegar) [inline Sources/Bow/Typeclasses/Applicative.swift]: I'd change this to the following, for consistency with the docs of `pure`:

```suggestion
/// An Applicative Functor is a `Functor` that also provides functionality to lift pure expressions, and sequence computations and combine their results.
```

File: Sources/Bow/Typeclasses/Functor.swift [hunk 2]
```diff
@@ -1,67 +1,166 @@
 import Foundation
 
+/// A Functor provides a type the ability to transform its value type into another type, while preserving its structure.
```
Comment (pepegar) [inline Sources/Bow/Typeclasses/Functor.swift]:
```suggestion
/// A Functor provides a type with the ability to transform its value type into another type, while preserving its structure.
```

File: Sources/Bow/Typeclasses/Monad.swift [hunk 3]
```diff
@@ -1,50 +1,135 @@
...
 public protocol Monad: Applicative {
+    /// Sequentially compose two computations, passing any value produced by the first as an argument to the second.
+    ///
+    /// - Parameters:
+    ///   - fa: First computation.
+    ///   - f: A function describing the second computation, which depends on the value of the first.
+    /// - Returns: Result of composing the two computations.
     static func flatMap<A, B>(_ fa: Kind<Self, A>, _ f: @escaping (A) -> Kind<Self, B>) -> Kind<Self, B>
+
+    /// Monadic tail recursion.
```
Comment (pepegar) [inline Sources/Bow/Typeclasses/Monad.swift]: In this comment I'd mention the reasoning behind this method and the paper that initially introduced it.

```suggestion
    /// Monadic tail recursion.
    ///
    /// tailRecM can be used for computations that can potentially make the stack overflow.
    ///
    /// initially introduced in Stack Safety for Free: https://functorial.com/stack-safety-for-free/index.pdf
```

Also, I'd mention that there's a default implementation for this, but it's not always lawful: (in scala, IDK any swift)

```scala
  def tailRecM[A, B](a: A)(f: A => F[Either[A, B]]): F[B] =
    flatMap(f(a)) {
      case Right(b) => pure(b)
      case Left(nextA) => tailRecM(nextA)(f)
    }
```

File: Sources/Bow/Typeclasses/Monad.swift [hunk 4]
```diff
@@ -1,50 +1,135 @@
...
+    /// - Parameters:
+    ///   - fa: 1st computation.
+    ///   - fb: 2nd computation.
+    /// - Returns: Result of running the second computation after the first one.
     public static func followedBy<A, B>(_ fa: Kind<Self, A>, _ fb: Kind<Self, B>) -> Kind<Self, B> {
         return self.flatMap(fa, { _ in fb })
     }
-    
+
+    /// Sequentially compose a computation with a lazy computation, discarding the value produced by the first.
```
Comment (pepegar) [inline Sources/Bow/Typeclasses/Monad.swift]: `Eval` is not always lazy, right? We could rephrase it as:

```suggestion
    /// Sequentially compose a computation with a potentially lazy one, discarding the value produced by the first.
```

File: Sources/Bow/Typeclasses/Monad.swift [hunk 5]
```diff
@@ -1,50 +1,135 @@
...
+    ///
+    /// - Parameters:
+    ///   - fa: Regular computation.
+    ///   - fb: Lazy computation.
```
Comment (pepegar) [inline Sources/Bow/Typeclasses/Monad.swift]:
```suggestion
    ///   - fb: Potentially lazy computation.
```

File: Sources/Bow/Typeclasses/Monad.swift [hunk 6]
```diff
@@ -1,50 +1,135 @@
...
+    /// - Parameters:
+    ///   - fa: 1st computation.
+    ///   - fb: 2nd computation.
+    /// - Returns: Result produced from the first computation after both are computed.
     public static func forEffect<A, B>(_ fa: Kind<Self, A>, _ fb: Kind<Self, B>) -> Kind<Self, A> {
         return self.flatMap(fa, { a in self.map(fb, { _ in a })})
     }
-    
+
+    /// Sequentially compose a computation with a lazy computation, discarding the value produced by the second.
```
Comment (pepegar) [inline Sources/Bow/Typeclasses/Monad.swift]: Same here. `Eval`'s not always lazy

File: Sources/Bow/Typeclasses/Monad.swift [hunk 7]
```diff
@@ -182,57 +342,160 @@ public extension Monad {
...
+    /// This is a convenience method to call `Monad.followedBy` as an instance method of this type.
+    ///
+    /// - Parameters:
+    ///   - fb: A computation.
+    /// - Returns: Result of running the second computation after the first one.
     public func followedBy<B>(_ fb: Kind<F, B>) -> Kind<F, B> {
         return F.followedBy(self, fb)
     }
 
+    /// Sequentially compose a computation with a lazy computation, discarding the value produced by this one.
```
Comment (pepegar) [inline Sources/Bow/Typeclasses/Monad.swift]:
```suggestion
    /// Sequentially compose a computation with a potentially lazy computation, discarding the value produced by this one.
```

File: Sources/Bow/Typeclasses/Monad.swift (hunk 3 above)
Comment (truizlop) [reply Sources/Bow/Typeclasses/Monad.swift]: We haven't made a default implementation, but it is something that I'd consider to add in the future.

**Comment by truizlop** [comment]
Thanks for your comments @pepegar! I just applied your comments.

**Pull request merged**